sql/core/src/main/scala/org/apache/spark/sql/streaming/ui/StreamingQueryStatisticsPage.scala

```diff
@@ -236,12 +239,61 @@ private[ui] class StreamingQueryStatisticsPage(parent: StreamingQueryTab)
           <td class={"aggregated-num-state-rows-dropped-by-watermark-timeline"}>{graphUIDataForNumRowsDroppedByWatermark.generateTimelineHtml(jsCollector)}</td>
           <td class={"aggregated-num-state-rows-dropped-by-watermark-histogram"}>{graphUIDataForNumRowsDroppedByWatermark.generateHistogramHtml(jsCollector)}</td>
         </tr>
-      // scalastyle:on
+        // scalastyle:on
+
+      result ++= generateAggregatedCustomMetrics(query, minBatchTime, maxBatchTime, jsCollector)
+      result
     } else {
       new NodeBuffer()
     }
   }
 
+  def generateAggregatedCustomMetrics(
+      query: StreamingQueryUIData,
+      minBatchTime: Long,
+      maxBatchTime: Long,
+      jsCollector: JsCollector): NodeBuffer = {
+    val result: NodeBuffer = new NodeBuffer
+
+    // This is made sure on caller side but put it here to be defensive
+    require(query.lastProgress.stateOperators.nonEmpty)
+    val enabledCustomMetrics = parent.parent.conf.get(ENABLED_STREAMING_UI_CUSTOM_METRIC_LIST)
+    logDebug(s"Enabled custom metrics: $enabledCustomMetrics")
+    query.lastProgress.stateOperators.head.customMetrics.keySet().asScala
+      .filter(enabledCustomMetrics.contains(_)).map { metricName =>
+        val data = query.recentProgress.map(p => (parseProgressTimestamp(p.timestamp),
+          p.stateOperators.map(_.customMetrics.get(metricName).toDouble).sum))
+        val max = data.maxBy(_._2)._2
+
+        val graphUIData =
+          new GraphUIData(
+            s"aggregated-$metricName-timeline",
+            s"aggregated-$metricName-histogram",
+            data,
+            minBatchTime,
+            maxBatchTime,
+            0,
+            max,
+            "")
```

Comment:
Hmm, though users might possibly infer the unit for the current available custom metrics,  we can't easily make the label of y-axis for custom metrics...

Reply:
It would be definitely good to add something better, but I have similar problem like in tooltip. Every custom metric can have its own unit.
I've just double checked and custom metrics stored in the following way: `java.util.Map[String, java.lang.Long]`
Since this structure doesn't contain any unit what the metric unit is I'm not sure from where it could come from.
Any suggestion?

Reply:
I think the way could be to introduce metric unit in the mentioned structure.
Since it doesn't exist I can't see possibilities for now.

Reply:
How about following idea?

1. Add `unit` to `StateStoreCustomMetric`
2. Call `StateStoreProvider.create(conf.get(STATE_STORE_PROVIDER_CLASS)).supportedCustomMetrics` and get the `Seq` of `StateStoreCustomMetric` from `StreamingQueryStatisticsPage`.
3. Refer `desc` and `unit` from the `StateStoreCustomMetric`

@HeartSaVioR  What do you tihnk?

Reply:
+1 on this idea, though not sure what kind of side effects we face. For instance does UI have `SparkEnv`.

Reply:
Never mind, these are `lazy val`s.

Reply:
The only concern what I see is that all `StateStoreProvider` implementation must support lightweight initialization which can be done from UI thread w/o exception. Not sure this implicit requirement is or will be known to all `StateStoreProvider` implementor.

Reply:
The idea looks promising. I see the concern from @gaborgsomogyi but given it has separate `init` method to bootstrap, more likely the implementation wouldn't do something in constructor.

Reply:
Since we're not calling `init` it doesn't play as you've mentioned. The restriction is not to put anything nasty in the constructor. I can live with it just wanted to highlight. Since I don't feel objection from @HeartSaVioR I'm starting to add it.

Reply:
I've added it and working fine so closing this thread.